File: plugins/harness/skills/harness-install/assets/uv/runtime/rules/require_hook_generated_marker.py
#!/usr/bin/env -S uv run
# -*- coding: utf-8 -*-
"""Require hook generated marker rule."""

from collections.abc import Iterable

from pathlib import Path

from harness_check_rule import Finding, HarnessCheckRule

from ._utils import is_safe_file, read_text, severity_for

class RequireHookGeneratedMarkerRule(HarnessCheckRule):
    """Validate requireHookGeneratedMarker check."""

    category = "requireHookGeneratedMarker"

    def applies(self, manifest: dict) -> bool:
        """Check if this rule applies to the manifest."""
        section = manifest.get(self.category)
        if not isinstance(section, dict):
            return False
        enabled = section.get("enabled", True)
        return enabled is not False

    def validate(self, root: Path, manifest: dict) -> Iterable[Finding]:
        """Validate requireHookGeneratedMarker check."""
        section = manifest.get(self.category, {})
        if not isinstance(section, dict):
            return []
        params = section.get("parameters", {})
        if not isinstance(params, dict):
            return []
        hooks = params.get("hooks", [])
        if not isinstance(hooks, list):
            return []
        marker_template = params.get("markerTemplate", "# Harness generated hook: {name}")
        placeholder_forbidden = params.get("placeholderForbidden", "packaged placeholder is replaced during harness installation")
        messages = section.get("messages", {})
        if not isinstance(messages, dict):
            return []
        valid_hooks = [
            (hook, read_text(root / hook), Path(hook).name)
            for hook in hooks
            if isinstance(hook, str) and is_safe_file(root / hook)
        ]
        findings = [
            Finding(
                severity_for(manifest, self.category),
                self.category,
                messages.get("missingMarker", "{hook} must contain generated marker '{marker}'").format(
                    hook=hook, marker=marker_template.format(name=hook_name)
                ),
            )
            for hook, text, hook_name in valid_hooks
            if marker_template.format(name=hook_name) not in text
        ]
        findings.extend(
            Finding(
                severity_for(manifest, self.category),
                self.category,
                messages.get("placeholderPresent", "{hook} still contains packaging placeholder text").format(hook=hook),
            )
            for hook, text, _ in valid_hooks
            if isinstance(placeholder_forbidden, str) and placeholder_forbidden in text
        )
        return findings


RULE: HarnessCheckRule = RequireHookGeneratedMarkerRule()

File: plugins/harness/skills/harness-install/assets/bun/runtime/rules/require-hook-generated-marker.ts
#!/usr/bin/env bun
import type {
	Finding,
	HarnessCheckRule,
	HarnessManifest,
	RuleContext,
} from "../harness-check-rule";

/**
 * Require hooks to contain generated markers.
 */
export const requireHookGeneratedMarkerRule = (ctx: RuleContext): HarnessCheckRule => ({
  applies(manifest: HarnessManifest): boolean {
    const section = manifest.requireHookGeneratedMarker;
    if (typeof section !== "object" || section === null) {
      return false;
    }
    const enabled = (section as { enabled?: unknown }).enabled;
    return enabled !== false && ctx.readStringArray(ctx.readJsonObject((section as Record<string, unknown>).parameters).hooks).length > 0;
  }

  validate(_root: string, manifest: HarnessManifest): readonly Finding[] {
    const parameters = ctx.readJsonObject(ctx.readJsonObject(manifest.requireHookGeneratedMarker).parameters);
    const hooks = ctx.readStringArray(parameters.hooks);
    const markerTemplate = typeof parameters.markerTemplate === "string" ? parameters.markerTemplate : "";
    const placeholderForbidden = typeof parameters.placeholderForbidden === "string" ? parameters.placeholderForbidden : "";
    return hooks.flatMap((hook) => {
      if (!ctx.isFile(hook)) {
        return [];
      }
      const marker = markerTemplate.replace("{name}", (hook.split("/").pop() ?? ""));
      const text = ctx.read(hook);
      return [
        !text.includes(marker)
          ? {
              severity: ctx.severityOf(manifest, "requireHookGeneratedMarker"),
              category: "requireHookGeneratedMarker",
              message: `${hook} must contain generated marker '${marker}'`,
            }
          : null,
        placeholderForbidden && text.includes(placeholderForbidden)
          ? {
              severity: ctx.severityOf(manifest, "requireHookGeneratedMarker"),
              category: "requireHookGeneratedMarker",
              message: `${hook} still contains packaging placeholder text`,
            }
          : null,
      ].filter((f): f is Finding => f !== null);
    });
  }

});
